From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Sun, 23 Jul 2023 11:11:53 +0200
Subject: [PATCH] Client perspective - Item replacements - Pass context of
 stonecutter recipe results

License: Fiddle Public License Agreement 1.0 (included in license/FPL-1.0.txt)
Fiddle - https://fiddlemc.org

diff --git a/src/main/java/net/minecraft/world/item/crafting/RecipeSerializer.java b/src/main/java/net/minecraft/world/item/crafting/RecipeSerializer.java
index c65ce69ac0cfdd04b16ce4ecd5b87536d727edd4..14bbd01eb15bb922d1562641cecf95892b8deed1 100644
--- a/src/main/java/net/minecraft/world/item/crafting/RecipeSerializer.java
+++ b/src/main/java/net/minecraft/world/item/crafting/RecipeSerializer.java
@@ -26,7 +26,7 @@ public interface RecipeSerializer<T extends Recipe<?>> {
     RecipeSerializer<BlastingRecipe> BLASTING_RECIPE = register("blasting", new SimpleCookingSerializer<>(BlastingRecipe::new, 100));
     RecipeSerializer<SmokingRecipe> SMOKING_RECIPE = register("smoking", new SimpleCookingSerializer<>(SmokingRecipe::new, 100));
     RecipeSerializer<CampfireCookingRecipe> CAMPFIRE_COOKING_RECIPE = register("campfire_cooking", new SimpleCookingSerializer<>(CampfireCookingRecipe::new, 100));
-    RecipeSerializer<StonecutterRecipe> STONECUTTER = register("stonecutting", new SingleItemRecipe.Serializer<>(StonecutterRecipe::new));
+    RecipeSerializer<StonecutterRecipe> STONECUTTER = register("stonecutting", new StonecutterRecipe.Serializer()); // Fiddle - client perspective - item replacements - pass context of stonecutter recipe results
     RecipeSerializer<SmithingTransformRecipe> SMITHING_TRANSFORM = register("smithing_transform", new SmithingTransformRecipe.Serializer());
     RecipeSerializer<SmithingTrimRecipe> SMITHING_TRIM = register("smithing_trim", new SmithingTrimRecipe.Serializer());
     RecipeSerializer<DecoratedPotRecipe> DECORATED_POT_RECIPE = register("crafting_decorated_pot", new SimpleCraftingRecipeSerializer<>(DecoratedPotRecipe::new));
diff --git a/src/main/java/net/minecraft/world/item/crafting/StonecutterRecipe.java b/src/main/java/net/minecraft/world/item/crafting/StonecutterRecipe.java
index 6d5803fcdd0a0b2d106745e9e6b859693b2aade2..2a9536867038f9a25a997d14e3688f06717c19c8 100644
--- a/src/main/java/net/minecraft/world/item/crafting/StonecutterRecipe.java
+++ b/src/main/java/net/minecraft/world/item/crafting/StonecutterRecipe.java
@@ -41,4 +41,29 @@ public class StonecutterRecipe extends SingleItemRecipe {
         return recipe;
     }
     // CraftBukkit end
+
+    // Fiddle start - client perspective - item replacements - pass context of stonecutter recipe results
+    /**
+     * A {@link SingleItemRecipe.Serializer} that writes the {@link #result}
+     * of a stone cutter recipe to a {@link net.minecraft.network.FriendlyByteBuf} with its
+     * {@linkplain org.fiddlemc.fiddle.packet.item.ItemStackInPacketContext#STONECUTTER_RECIPE_RESULT distinguishing context}.
+     */
+    public static class Serializer extends SingleItemRecipe.Serializer<StonecutterRecipe> {
+
+        protected Serializer() {
+            super(StonecutterRecipe::new);
+        }
+
+        @Override
+        public void toNetwork(net.minecraft.network.FriendlyByteBuf buf, StonecutterRecipe recipe) {
+            // Based on SingleItemRecipe.Serializer<StonecutterRecipe>#toNetwork
+            buf.writeUtf(recipe.group);
+            recipe.ingredient.toNetwork(buf);
+            // Use the distinguishing context
+            buf.writeItem(recipe.result, org.fiddlemc.fiddle.packet.item.ItemStackInPacketContext.STONECUTTER_RECIPE_RESULT);
+        }
+
+    }
+    // Fiddle end - client perspective - item replacements - pass context of stonecutter recipe results
+
 }
